---
title: "Differential Gene Expression Analysis"
author: "Wenjing Meng"
date: "`r Sys.Date()`"
output: html_document
---
```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```
# 1Count normalization
## 1.1Install packages
Install DESeq2 package
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```
install EdgeR package because we need the readDGE function
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

## 1.2Read data files
Set to the current work directory
```{r}
setwd("D:/PHD/Courses/Practical Course in Boinformatics/DIferentialGeneExpressionAnalysis/RNA_seq_DEA")
```
read raw data
```{r}
Coinfection.targets<-read.delim("./data/fileDesc.txt")
```
change the rawnames of the dataframe 
```{r}
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
```
load the packege edgeR
```{r}
library(edgeR)
```
read the six .txt files listed in Coinfection.targets
```{r}
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
```
check the dimension of the data set
```{r}
dim(Coinfection.orig)
```
check the first 6 rows of the data
```{r}
head(Coinfection.orig)
```
Extract counts dataframe
```{r}
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)
```
```{r}
head(Coinfection.rawCount)
```
## 1.3Build meta data
```{r}
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))
```
Build meta data frame
```{r}
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))
```
Check the column name of counts dataframe
```{r}
colnames(Coinfection.orig$count)
```
Check the rowname of meta dataframe
```{r}
rownames(meta)
```
Check that sample names match in both files
```{r}
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```
## 1.4 Create DESeqDataset object
```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)
```
## 1.5 Generate size factors
To perform the median of ratios method of normalization, DESeq2 has a single estimateSizeFactors() function that will generate size factors.
```{r}
dds <- estimateSizeFactors(dds)
```
Look at the normalization factor applied to each sample
```{r}
sizeFactors(dds)
```
## 1.6 Generate the normalized counts
Retrieve the normalized counts matrix from dds
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
```
Save the normalized_counts
```{r}
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```

# 2 Sample-level quality Control
## 2.1 Principal Component Analysis (PCA)
Transform counts for data visualization
```{r}
rld <- rlog(dds, blind=TRUE)
```
 **start with PCA*
```{r}
plotPCA(rld, intgroup="sampletype")
```
save PDF file
```{r}
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()
```
**Exercise**:
1.What does the above plot tell you about the similarity of samples?
2.Does it fit the expectation from the experimental design?
3.What do you think the %variance information (in the axes titles) tell you about the data in the context of the PCA?

1.The control group and HA-infected group seperated significantly.
2.Yes.
3.PC1 captures half of the patterns or differences present in the data, making it the most significant axis in terms of explaining the variance.PC2 explains 25% of the total variance, so together the two axes explained 75% variation.

##2.2 Hierarchical Clustering Heatmap
Extract the rlog matrix from the object
```{r}
rld_mat <- assay(rld)
```
Compute the pairwise correlation values for all the samples
```{r}
rld_cor <- cor(rld_mat) 
```
Check the output of cor(), make note of the row names and column names
```{r}
head(rld_cor)
```
Check meta
```{r}
head(meta)
```
Install pheatmap package and load the package
```{r}
install.packages("pheatmap")
```
Plot heatmap using the correlation matrix and the metadata object
```{r}
library(pheatmap)
pheatmap(rld_cor, annotation = meta)
```
Change colors
```{r}
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
        fontsize_row = 10, height=20)
```
Out put PDF file
```{r}
pdf("./results/PlotHeatmap_dds.pdf")
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
        fontsize_row = 10, height=20)
```

# 3 Differential expression analysis (DEA) using EdgeR
load EdgeR package
```{r}
library(edgeR)
options(digits=3)
```
Tell R where the data files are
```{r}
infection.targets<-read.delim("./data/fileDesc.txt")
```
Check Coinfection.targets
```{r}
infection.targets
```
Change the rawnames of the dataframe Coinfection.targets, 
```{r}
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
```
Check Coinfection.targets again
```{r}
infection.targets
```
read and merges a set of text files containing gene expression counts, it makes a DGEList object directly
```{r}
infection <- readDGE(infection.targets, header=F)
```
Check the dimension of DGElist R object
```{r}
dim(infection)
```
```{r}
head(infection)
```
Get the raw mapped count before filtering
```{r}
infection.rawCount <- infection$count
head(infection.rawCount)
```
```{r}
library(ggplot2)
```
To get an idea about how RNA-seq counts are distributed,  plot a histogram of the counts for a single sample, ‘Ha1’:
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
```
Export the .png file
```{r}
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```
Export raw count table into results folder
```{r}
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")
```
Get the counts per million (TMM normalised) before filtering
```{r}
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)
```
```{r}
head(infection.normCPM)
```
```{r}
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
```
Keep genes that are expressed at least 1 CPM in at least 3 libraries, normally it is the number of biological replicates of smaller group
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3
```
```{r}
table(infection.filtered)
```
Libraries size of data BEFORE filtering
```{r}
infection$samples$lib.size
```
cover the original file with our filter data
```{r}
Infection <- infection[infection.filtered,]
```
libraries size of data after filtering
```{r}
colSums(Infection$counts)
```
```{r}
dim(Infection)
```
Update the filtered libraries size
```{r}
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
```
Perform normalization with TMM method
```{r}
Infection = calcNormFactors(Infection)
```
The libraries after normalisation
```{r}
Infection$samples
```
Get the counts per million (TMM normalised) after filtering
```{r}
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
```
Export TMM normalized count table after filtering
```{r}
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```
## differentailly gene expression analysis

```{r}
getwd()
```
treatment factor
```{r}
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
```
Describe the experimental design,one factor with intercept, here Ctr is the intercept
```{r}
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
```
To see if some samples are outliner. An MDS plot shows the relative similarities of the six samples
```{r}
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
```
Reset the working path because I changed the directory files to link with github
```{r}
setwd("D:/PHD/Courses/Practical Course in Boinformatics/DIferentialGeneExpressionAnalysis/RNA_seq_DEA-")
```

```{r}
library(edgeR)
```
Export the MDS plot to a PNG file
```{r}
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()
```
Estimating the common dispersion, trended dispersion, tagwwise dispersion Dispersion means biological coeffient of variation (BCV) squared. E.g. if genes expression typically differs from replicate to replicate by 20% its BCV is 0.2, and its dispersion is 0.04. 
Estimating the common dispersion
```{r}
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
```
Estimating the trended dispersion
```{r}
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
```
Estimating the tagwwise dispersion
```{r}
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)
```
Creating a visual representation of the mean-variance relationship and BCV-mean logCPM relationship
```{r}
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
```
```{r}
plotBCV(Infection)
```
Fit DGEList and design matrix to genewise GLM
```{r}
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
```
Likelihood test for genes respond differently between different conditions, table of top differentially expressed tags, n specify n tags to display
```{r}
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)  # to compare Ha vs Ctr (Ha_vs_Ctr)
```
```{r}
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))
head(t1$table)
```
Total number of differentially expressed (DE) genes in different comparsion at FDR < 0.05
```{r}
summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))
```
The UP-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
```
```{r}
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
```
The DW-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))
```
```{r}
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)
```
Differentially expressed transcripts’ tag
```{r}
DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]
```
Export list of UP-regulated and DW-regulated transcripts
```{r}
write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")
```
To ask all the genes label as grey color
```{r}
Infection.colHavsCtr = rep('grey55', nrow(Infection))
```
To assign colour for DE transcripts
```{r}
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"
```
Plot Smear plot with adjustment on Y-axis label
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")
```
```{r}
abline(h=c(-1,1),col="dodgerblue")
```


```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="a infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
```

```{r}
abline(h=c(-1,1),col="dodgerblue")
```
Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed and export as png
```{r}
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
dev.off()
```
